File: windows/AI/Data/AnalysisStreaming.cs
using System.Runtime.CompilerServices;
using System.Text.Json;

namespace AI.Data;

public static class AnalysisStreaming
{
    private static readonly AnalysisRepository Repo = AnalysisRepository.Instance;

    public static IAsyncEnumerable<string> SendChatMessageStream(
        AppService service, string apiKey, string model, List<ChatMessage> messages,
        ChatParameters parms, string? customBaseUrl = null, CancellationToken ct = default)
    {
        return service.ApiFormat switch
        {
            ApiFormat.Anthropic => StreamChatClaude(service, apiKey, model, messages, parms, ct),
            ApiFormat.Google => StreamChatGemini(service, apiKey, model, messages, parms, ct),
            _ => StreamChatOpenAiCompatible(service, apiKey, model, messages, parms, customBaseUrl, ct)
        };
    }

    private static async IAsyncEnumerable<string> StreamChatOpenAiCompatible(
        AppService service, string apiKey, string model, List<ChatMessage> messages,
        ChatParameters parms, string? customBaseUrl, [EnumeratorCancellation] CancellationToken ct)
    {
        if (Repo.UsesResponsesApi(service, model))
        {
            await foreach (var chunk in StreamResponsesApi(service, apiKey, model, messages, parms, customBaseUrl, ct))
                yield return chunk;
            yield break;
        }

        var openAiMessages = Repo.ConvertToOpenAiMessages(messages, parms.SystemPrompt);
        var request = new OpenAiStreamRequest
        {
            Model = model,
            Messages = openAiMessages,
            Stream = true,
            MaxTokens = parms.MaxTokens,
            Temperature = parms.Temperature,
            TopP = parms.TopP,
            TopK = parms.TopK,
            FrequencyPenalty = parms.FrequencyPenalty,
            PresencePenalty = parms.PresencePenalty,
            ReturnCitations = service.SupportsCitations ? parms.ReturnCitations : null,
            SearchRecencyFilter = service.SupportsSearchRecency ? parms.SearchRecency : null,
            Search = parms.SearchEnabled ? true : null
        };

        var url = ApiClient.ChatUrl(service, customBaseUrl);
        using var response = await ApiClient.Instance.StreamRequestAsync(url, request, service, apiKey);
        using var stream = await response.Content.ReadAsStreamAsync(ct);
        using var reader = new StreamReader(stream);

        while (!reader.EndOfStream && !ct.IsCancellationRequested)
        {
            var line = await reader.ReadLineAsync(ct);
            if (line == null) break;
            if (!line.StartsWith("data: ")) continue;

            var data = line[6..];
            if (data == "[DONE]") break;

            OpenAiStreamChunk? chunk;
            try { chunk = JsonSerializer.Deserialize<OpenAiStreamChunk>(data); }
            catch { continue; }

            if (chunk?.Choices?.FirstOrDefault()?.Delta?.Content is { } content)
                yield return content;
            if (chunk?.Choices?.FirstOrDefault()?.Delta?.ReasoningContent is { } reasoning)
                yield return reasoning;
        }
    }

    private static async IAsyncEnumerable<string> StreamResponsesApi(
        AppService service, string apiKey, string model, List<ChatMessage> messages,
        ChatParameters parms, string? customBaseUrl, [EnumeratorCancellation] CancellationToken ct)
    {
        var inputMessages = messages.Where(m => m.Role != "system")
            .Select(m => new OpenAiResponsesInputMessage { Role = m.Role, Content = m.Content }).ToList();

        var request = new OpenAiResponsesStreamRequest
        {
            Model = model,
            Input = inputMessages,
            Instructions = string.IsNullOrEmpty(parms.SystemPrompt) ? null : parms.SystemPrompt,
            Stream = true
        };

        var baseUrl = customBaseUrl ?? service.BaseUrl;
        var normalizedBase = baseUrl.EndsWith("/") ? baseUrl : $"{baseUrl}/";
        var url = $"{normalizedBase}v1/responses";

        using var response = await ApiClient.Instance.StreamRequestAsync(url, request, service, apiKey);
        using var stream = await response.Content.ReadAsStreamAsync(ct);
        using var reader = new StreamReader(stream);

        while (!reader.EndOfStream && !ct.IsCancellationRequested)
        {
            var line = await reader.ReadLineAsync(ct);
            if (line == null) break;
            if (!line.StartsWith("data: ")) continue;

            var data = line[6..];
            if (data == "[DONE]") break;

            try
            {
                using var doc = JsonDocument.Parse(data);
                if (doc.RootElement.TryGetProperty("delta", out var delta) && delta.ValueKind == JsonValueKind.String)
                    yield return delta.GetString()!;
            }
            catch { continue; }
        }
    }

    private static async IAsyncEnumerable<string> StreamChatClaude(
        AppService service, string apiKey, string model, List<ChatMessage> messages,
        ChatParameters parms, [EnumeratorCancellation] CancellationToken ct)
    {
        var claudeMessages = messages.Where(m => m.Role != "system")
            .Select(m => new ClaudeMessage { Role = m.Role == "assistant" ? "assistant" : "user", Content = m.Content }).ToList();

        var request = new ClaudeStreamRequest
        {
            Model = model,
            Messages = claudeMessages,
            Stream = true,
            MaxTokens = parms.MaxTokens ?? 4096,
            Temperature = parms.Temperature,
            TopP = parms.TopP,
            TopK = parms.TopK,
            System = string.IsNullOrEmpty(parms.SystemPrompt) ? null : parms.SystemPrompt,
            Search = parms.SearchEnabled ? true : null
        };

        var baseUrl = service.BaseUrl.EndsWith("/") ? service.BaseUrl : $"{service.BaseUrl}/";
        var url = $"{baseUrl}v1/messages";

        using var response = await ApiClient.Instance.StreamRequestAsync(url, request, service, apiKey);
        using var stream = await response.Content.ReadAsStreamAsync(ct);
        using var reader = new StreamReader(stream);

        var currentEvent = "";
        while (!reader.EndOfStream && !ct.IsCancellationRequested)
        {
            var line = await reader.ReadLineAsync(ct);
            if (line == null) break;

            if (line.StartsWith("event: "))
            {
                currentEvent = line[7..];
                if (currentEvent == "message_stop") break;
            }
            else if (line.StartsWith("data: ") && currentEvent == "content_block_delta")
            {
                var data = line[6..];
                try
                {
                    var evt = JsonSerializer.Deserialize<ClaudeStreamEvent>(data);
                    if (evt?.Delta?.Text is { } text)
                        yield return text;
                }
                catch { continue; }
            }
        }
    }

    private static async IAsyncEnumerable<string> StreamChatGemini(
        AppService service, string apiKey, string model, List<ChatMessage> messages,
        ChatParameters parms, [EnumeratorCancellation] CancellationToken ct)
    {
        var geminiMessages = messages.Where(m => m.Role != "system")
            .Select(m => new GeminiContent
            {
                Parts = new() { new GeminiPart { Text = m.Content } },
                Role = m.Role == "assistant" ? "model" : "user"
            }).ToList();

        var config = new GeminiGenerationConfig
        {
            Temperature = parms.Temperature,
            TopP = parms.TopP,
            TopK = parms.TopK,
            MaxOutputTokens = parms.MaxTokens
        };

        var systemInstruction = string.IsNullOrEmpty(parms.SystemPrompt) ? null
            : new GeminiContent { Parts = new() { new GeminiPart { Text = parms.SystemPrompt } } };

        var request = new GeminiRequest
        {
            Contents = geminiMessages,
            GenerationConfig = config,
            SystemInstruction = systemInstruction
        };

        var url = ApiClient.GeminiGenerateUrl(service.BaseUrl, model, apiKey, stream: true);
        var dummyService = new AppService { Id = "gemini_temp", ApiFormatStr = "GOOGLE" };

        using var response = await ApiClient.Instance.StreamRequestAsync(url, request, dummyService, apiKey);
        using var stream = await response.Content.ReadAsStreamAsync(ct);
        using var reader = new StreamReader(stream);

        while (!reader.EndOfStream && !ct.IsCancellationRequested)
        {
            var line = await reader.ReadLineAsync(ct);
            if (line == null) break;
            if (!line.StartsWith("data: ")) continue;

            var data = line[6..];
            try
            {
                var chunk = JsonSerializer.Deserialize<GeminiStreamChunk>(data);
                if (chunk?.Candidates?.FirstOrDefault()?.Content?.Parts?.FirstOrDefault()?.Text is { } text)
                    yield return text;
            }
            catch { continue; }
        }
    }
}

File: mac/macAI/Data/AnalysisStreaming.swift
import Foundation
import os

/// SSE streaming parsers -> AsyncThrowingStream.
extension AnalysisRepository {

    // MARK: - Stream Entry Point

    func sendChatMessageStream(
        service: AppService,
        apiKey: String,
        model: String,
        messages: [ChatMessage],
        params: ChatParameters,
        customBaseUrl: String? = nil
    ) -> AsyncThrowingStream<String, Error> {
        switch service.apiFormat {
        case .openaiCompatible:
            return streamChatOpenAiCompatible(service: service, apiKey: apiKey, model: model, messages: messages, params: params, customBaseUrl: customBaseUrl)
        case .anthropic:
            return streamChatClaude(service: service, apiKey: apiKey, model: model, messages: messages, params: params)
        case .google:
            return streamChatGemini(service: service, apiKey: apiKey, model: model, messages: messages, params: params)
        }
    }

    // MARK: - OpenAI Compatible Streaming

    private func streamChatOpenAiCompatible(
        service: AppService,
        apiKey: String,
        model: String,
        messages: [ChatMessage],
        params: ChatParameters,
        customBaseUrl: String? = nil
    ) -> AsyncThrowingStream<String, Error> {
        // Check for Responses API
        if usesResponsesApi(service, model: model) {
            return streamResponsesApi(service: service, apiKey: apiKey, model: model, messages: messages, params: params, customBaseUrl: customBaseUrl)
        }

        let openAiMessages = convertToOpenAiMessages(messages, systemPrompt: params.systemPrompt)

        let request = OpenAiStreamRequest(
            model: model,
            messages: openAiMessages,
            stream: true,
            max_tokens: params.maxTokens,
            temperature: params.temperature,
            top_p: params.topP,
            top_k: params.topK,
            frequency_penalty: params.frequencyPenalty,
            presence_penalty: params.presencePenalty,
            seed: service.seedFieldName == "seed" ? nil : nil,  // Seed not typically used in chat streaming
            return_citations: service.supportsCitations ? params.returnCitations : nil,
            search_recency_filter: service.supportsSearchRecency ? params.searchRecency : nil,
            search: params.searchEnabled ? true : nil
        )

        let url = ApiClient.chatUrl(for: service, customBaseUrl: customBaseUrl)

        return AsyncThrowingStream { continuation in
            Task {
                do {
                    let (bytes, _) = try await ApiClient.shared.streamRequest(url: url, body: request, service: service, apiKey: apiKey)
                    for try await line in bytes.lines {
                        if line.hasPrefix("data: ") {
                            let data = String(line.dropFirst(6))
                            if data == "[DONE]" { break }
                            if let jsonData = data.data(using: .utf8),
                               let chunk = try? JSONDecoder().decode(OpenAiStreamChunk.self, from: jsonData) {
                                if let content = chunk.choices?.first?.delta?.content {
                                    continuation.yield(content)
                                }
                                if let reasoning = chunk.choices?.first?.delta?.reasoning_content {
                                    continuation.yield(reasoning)
                                }
                            }
                        }
                    }
                    continuation.finish()
                } catch {
                    continuation.finish(throwing: error)
                }
            }
        }
    }

    // MARK: - Responses API Streaming

    private func streamResponsesApi(
        service: AppService,
        apiKey: String,
        model: String,
        messages: [ChatMessage],
        params: ChatParameters,
        customBaseUrl: String? = nil
    ) -> AsyncThrowingStream<String, Error> {
        let inputMessages = messages.filter { $0.role != "system" }.map {
            OpenAiResponsesInputMessage(role: $0.role, content: $0.content)
        }

        let request = OpenAiResponsesStreamRequest(
            model: model,
            input: inputMessages,
            instructions: params.systemPrompt.isEmpty ? nil : params.systemPrompt,
            stream: true
        )

        let base = customBaseUrl ?? service.baseUrl
        let normalizedBase = base.hasSuffix("/") ? base : "\(base)/"
        let url = "\(normalizedBase)v1/responses"

        return AsyncThrowingStream { continuation in
            Task {
                do {
                    let (bytes, _) = try await ApiClient.shared.streamRequest(url: url, body: request, service: service, apiKey: apiKey)
                    for try await line in bytes.lines {
                        if line.hasPrefix("data: ") {
                            let data = String(line.dropFirst(6))
                            if data == "[DONE]" { break }
                            if let jsonData = data.data(using: .utf8),
                               let json = try? JSONSerialization.jsonObject(with: jsonData) as? [String: Any] {
                                // Look for delta text in response.output_text.delta events
                                if let delta = json["delta"] as? String {
                                    continuation.yield(delta)
                                }
                            }
                        }
                    }
                    continuation.finish()
                } catch {
                    continuation.finish(throwing: error)
                }
            }
        }
    }

    // MARK: - Claude Streaming

    private func streamChatClaude(
        service: AppService,
        apiKey: String,
        model: String,
        messages: [ChatMessage],
        params: ChatParameters
    ) -> AsyncThrowingStream<String, Error> {
        let claudeMessages = messages.filter { $0.role != "system" }.map {
            ClaudeMessage(role: $0.role == "assistant" ? "assistant" : "user", content: $0.content)
        }

        let request = ClaudeStreamRequest(
            model: model,
            messages: claudeMessages,
            stream: true,
            max_tokens: params.maxTokens ?? 4096,
            temperature: params.temperature,
            top_p: params.topP,
            top_k: params.topK,
            system: params.systemPrompt.isEmpty ? nil : params.systemPrompt,
            search: params.searchEnabled ? true : nil
        )

        let base = service.baseUrl.hasSuffix("/") ? service.baseUrl : "\(service.baseUrl)/"
        let url = "\(base)v1/messages"

        return AsyncThrowingStream { continuation in
            Task {
                do {
                    let (bytes, _) = try await ApiClient.shared.streamRequest(url: url, body: request, service: service, apiKey: apiKey)
                    var currentEvent = ""
                    for try await line in bytes.lines {
                        if line.hasPrefix("event: ") {
                            currentEvent = String(line.dropFirst(7))
                            if currentEvent == "message_stop" { break }
                        } else if line.hasPrefix("data: "), currentEvent == "content_block_delta" {
                            let data = String(line.dropFirst(6))
                            if let jsonData = data.data(using: .utf8),
                               let event = try? JSONDecoder().decode(ClaudeStreamEvent.self, from: jsonData),
                               let text = event.delta?.text {
                                continuation.yield(text)
                            }
                        }
                    }
                    continuation.finish()
                } catch {
                    continuation.finish(throwing: error)
                }
            }
        }
    }

    // MARK: - Gemini Streaming

    private func streamChatGemini(
        service: AppService,
        apiKey: String,
        model: String,
        messages: [ChatMessage],
        params: ChatParameters
    ) -> AsyncThrowingStream<String, Error> {
        let geminiMessages = messages.filter { $0.role != "system" }.map {
            GeminiContent(parts: [GeminiPart(text: $0.content)], role: $0.role == "assistant" ? "model" : "user")
        }

        let config = GeminiGenerationConfig(
            temperature: params.temperature,
            topP: params.topP,
            topK: params.topK,
            maxOutputTokens: params.maxTokens
        )

        let systemInstruction = params.systemPrompt.isEmpty ? nil : GeminiContent(parts: [GeminiPart(text: params.systemPrompt)])

        let request = GeminiRequest(
            contents: geminiMessages,
            generationConfig: config,
            systemInstruction: systemInstruction
        )

        let url = ApiClient.geminiGenerateUrl(baseUrl: service.baseUrl, model: model, apiKey: apiKey, stream: true)

        return AsyncThrowingStream { continuation in
            Task {
                do {
                    // Use a dummy service since Gemini auth is in URL
                    let dummyService = AppService(id: "gemini_temp", displayName: "", baseUrl: "", defaultModel: "", apiFormat: .google)
                    let (bytes, _) = try await ApiClient.shared.streamRequest(url: url, body: request, service: dummyService, apiKey: apiKey)
                    for try await line in bytes.lines {
                        if line.hasPrefix("data: ") {
                            let data = String(line.dropFirst(6))
                            if let jsonData = data.data(using: .utf8),
                               let chunk = try? JSONDecoder().decode(GeminiStreamChunk.self, from: jsonData),
                               let text = chunk.candidates?.first?.content?.parts.first?.text {
                                continuation.yield(text)
                            }
                        }
                    }
                    continuation.finish()
                } catch {
                    continuation.finish(throwing: error)
                }
            }
        }
    }

    // MARK: - Message Conversion

    func convertToOpenAiMessages(_ messages: [ChatMessage], systemPrompt: String? = nil) -> [OpenAiMessage] {
        var result: [OpenAiMessage] = []
        if let sys = systemPrompt, !sys.isEmpty {
            result.append(OpenAiMessage(role: "system", content: sys))
        }
        result += messages.map { OpenAiMessage(role: $0.role, content: $0.content) }
        return result
    }
}
